misk-crypto/src/main/kotlin/misk/crypto/pgp/RealPgpEncoder.kt

```diff
@@ -0,0 +1,133 @@
+import org.bouncycastle.openpgp.PGPCompressedDataGenerator
+import org.bouncycastle.openpgp.PGPEncryptedData
+import org.bouncycastle.openpgp.PGPEncryptedDataGenerator
+import org.bouncycastle.openpgp.PGPEncryptedDataList
+import org.bouncycastle.openpgp.PGPException
+import org.bouncycastle.openpgp.PGPLiteralData
+import org.bouncycastle.openpgp.PGPLiteralDataGenerator
+import org.bouncycastle.openpgp.PGPOnePassSignatureList
+import org.bouncycastle.openpgp.PGPPrivateKey
+import org.bouncycastle.openpgp.PGPPublicKey
+import org.bouncycastle.openpgp.PGPPublicKeyEncryptedData
+import org.bouncycastle.openpgp.PGPUtil
+import org.bouncycastle.openpgp.jcajce.JcaPGPObjectFactory
+import org.bouncycastle.openpgp.operator.jcajce.JcePGPDataEncryptorBuilder
+import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyDataDecryptorFactoryBuilder
+import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyKeyEncryptionMethodGenerator
+import org.bouncycastle.util.io.Streams
+import java.io.ByteArrayOutputStream
+import java.security.SecureRandom
+import java.security.Security
+import java.util.Date
+
+internal class RealPgpEncoder(
+  private val privateKey: PGPPrivateKey
+) : PgpEncoder {
+  init {
+    Security.addProvider(BouncyCastleProvider())
+  }
+
+  override fun encrypt(
+    byteArray: ByteArray,
+    encryptionKey: PGPPublicKey,
+    armored: Boolean
+  ): ByteArray {
+    val buffer = okio.Buffer()
+
+    val byteArrayOutputStream = ByteArrayOutputStream()
+    val pgpCompressedDataGenerator = PGPCompressedDataGenerator(PGPCompressedData.ZIP)
+    val pgpLiteralDataGenerator = PGPLiteralDataGenerator()
+    val pgpLiteralDataStream = pgpLiteralDataGenerator.open(
+        pgpCompressedDataGenerator.open(byteArrayOutputStream),
+        PGPLiteralData.BINARY,
+        "", // Name of the file encoded in the data. Just boring metadata.
+        byteArray.size.toLong(),
+        Date()
+    )
+
+    pgpLiteralDataStream.write(byteArray)
+    pgpLiteralDataStream.close()
+    pgpCompressedDataGenerator.close()
+
+    val outputStream = buffer.outputStream()
+    val out = when (armored) {
+      true -> ArmoredOutputStream(outputStream)
+      else -> outputStream
+    }
+
+    val jcePGPDataEncryptorBuilder = JcePGPDataEncryptorBuilder(PGPEncryptedData.CAST5)
+        .setWithIntegrityPacket(true)
+        .setSecureRandom(SecureRandom())
+        .setProvider("BC")
+
+    val pgpEncryptedDataGenerator = PGPEncryptedDataGenerator(jcePGPDataEncryptorBuilder)
+    val keyEncryptionMethodGenerator = JcePublicKeyKeyEncryptionMethodGenerator(encryptionKey)
+        .setProvider(BouncyCastleProvider())
+        .setSecureRandom(SecureRandom())
+    pgpEncryptedDataGenerator.addMethod(keyEncryptionMethodGenerator)
+
+    val bytes = byteArrayOutputStream.toByteArray()
+    val encryptedOut = pgpEncryptedDataGenerator.open(out, bytes.size.toLong())
+    encryptedOut.write(bytes)
+    encryptedOut.close()
+
+    if (armored) {
```

Comment:
What does armored mean in this context?

Reply:
armored is a PGP specific term for whether we use a binary output or an emailable-text form like ```-----BEGIN PGP MESSAGE-----
Version: BCPG C# v1.6.1.0

hQIMA0tM4ZUzpKCkAQ//ZO8hVhp6LMxshCLqcqgPkXcU1kBHPQjiUBs6QPxaQFcg
PIXJEdZWQ1RkMUKdCfzjvevizi/2NWAwZ797fELdbpVyH6JI2Me1Ov10f8qgYvbG
V0GJ3loRt84KjKdVocKguJVfMvOeMVJanPpxfPt3Ak+Nt/voZHmJdTkV4cjWTmcU
x31xqc5LZcvgF4Lv+ZcMw090b281yMENwALnvQJ9FQg5WTfZ+YautUMwWGk8W8XM
H3tYkdh3jKizy8YTw+zfVDn1yhw2BgSFp0QIgHIQQ2Q2gyknRfj4zdfs/4bZYcu1
KNgj9op/IejZNstP9JWw4labC78nTIW5s1f0LWF5hhZUkxeAOEyVluUVJgEw44B4
LKb0p/8jb4J/dVrsmaTDOONZ48wtNjZOTV61DhCnn+UiNVSfQe6I7XMtFpEH0PN4
WzermmeEgjw1MADeh0jg/9wMM/p9UnTxihGQrPRgHYUu45jI3ys3qArsD2GvAi9y
QYqsuRJNB7EEu2clBRASA3zoWkpqbs3H9S2fB9Sf9ZzZTPCzowOdK1oe0vlMVt3A
mCILe05s2nowwBrBWu4PHUTkllsnLliHhrihu9sqXCHy6u1pdNQMvJgY6kbgQmsG
bVY+brdkLdQ1BxbTT1YixzLm//zqMqGC1ZJUsr5PAIbNtyKK6u6+mCkZj0Y3fU3J
Il0x5V09QcC9AHyxJ8VMOxzhQZjwVYNtD6SGeY8jOyfUScY=
=DHKq
-----END PGP MESSAGE-----
```

Reply:
👍